src/System.Text.Json/src/System/Text/Json/Serialization/CachedUtf8JsonWriter.cs

```diff
@@ -3,56 +3,26 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Buffers;
+using System.Collections.Concurrent;
 
 namespace System.Text.Json.Serialization
 {
     internal sealed class CachedUtf8JsonWriter
     {
-        [ThreadStatic]
-        private static CachedUtf8JsonWriter s_cachedInstance;
+        private static readonly ConcurrentDictionary<int, Utf8JsonWriter> _cache = new ConcurrentDictionary<int, Utf8JsonWriter>();
 
-        private readonly Utf8JsonWriter _writer;
-
-#if DEBUG
-        private bool _inUse;
-#endif
-
-        private CachedUtf8JsonWriter(IBufferWriter<byte> output, JsonWriterOptions writerOptions)
-        {
-            _writer = new Utf8JsonWriter(output, writerOptions);
-        }
-
-        public static CachedUtf8JsonWriter Get(IBufferWriter<byte> output, JsonWriterOptions writerOptions)
+        public static Utf8JsonWriter Get(IBufferWriter<byte> output, JsonWriterOptions writerOptions)
         {
-            var writer = s_cachedInstance ?? new CachedUtf8JsonWriter(output, writerOptions);
+            int writerOptionsHash = writerOptions.GetHashCode();
 
-            // Taken off the thread static
-            s_cachedInstance = null;
-#if DEBUG
-            if (writer._inUse)
+            if (!_cache.TryGetValue(writerOptionsHash, out Utf8JsonWriter writer))
```

Comment:
I don't think this is thread safe anymore. Two threads with the same options could use the same writer

Reply:
Can you measure perf. It may be more expensive accessing from a dictionary than newing it up each time even considering gc

Reply:
@steveharter [see](https://github.com/dotnet/corefx/pull/37533#discussion_r283114518)

Reply:
>Two threads with the same options could use the same writer

Yes if two threads can use the same instance then they would share the same buffer and internal writer state which would be bad.

Reply:
In that case, maybe try putting it under a lock and see how it affects perf?